sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsCheckpointMark.java

```diff
@@ -59,12 +63,15 @@ public void finalizeCheckpoint() {
     for (Message message : messages) {
       try {
         message.acknowledge();
+        Instant currentMessageTimestamp = new Instant(message.getJMSTimestamp());
+        if (currentMessageTimestamp.isAfter(oldestPendingTimestamp)) {
```

Comment:
Add a comment here, something like:

```
// Acknowledge all outstanding messages. Since we believe that messages
// will be delivered in timestamp order, and acknowledged messages will not be retried,
// the newest message in this batch is a good lower bound for future messages.
```

(Could also use oldest message for a more conservative bound in case some messages do get retried anyway, but that's up to you.)